src/rsdiv/evaluation/diversity_metrics.py

```diff
@@ -0,0 +1,19 @@
+import numpy as np
+
+
+def gini_coefficient(x: np.ndarray, sort: bool = True) -> float:
+    if sort:
+        x = np.sort(x)[::-1]
+    count: int = x.shape[0]
+    b: float = x @ np.arange(1, count + 1)
+    b /= x.sum() * count
+    return 1 - 2 * b + 1 / count
+
+
+def effective_catalog_size(x: np.ndarray, sort: bool = True) -> float:
```

Comment:
try to make function compatible to the usage as follows:
```
data_loader = MovieLens1MDownLoader()
ratings = downloader.read_ratings()
movies = downloader.read_movies()

metrics = DiversityMetrics()
gini_ratings = metrics.gini_coefficient(ratings['movieId'])
gini_movies = metrics.gini_coefficient(movies['genres'])
```
caution `ratings['movieId']` is a kind of List[str], and `movies['genres']` is a kind of `List[List[str]]`

Reply:
Is `ratings['movieId']` supposed to be `pd.Series[int]`?
And you also mentioned that method signatures should be
```
class DiversityMetrics():

@classmethod
def gini_coefficient(cls, items: pd.Series[str], sort: bool = True) -> float:

@classmethod
def effective_catalog_size(cls, items: pd.Series[str], sort: bool = True) -> float:

```
Which one should I use?

Reply:
we can just use typehint as `pd.Series`, If data type is not specified, it will be inferred from data.

Reply:
```
items: pandas.Series
```

Reply:
When an item have multiple categories (i.e. `items` is `pd.Series[List]`), should this item be considered as multiple items when computing metrics?

Reply:
Yes. For the case of `movies['genres']`, it is a style of `List[List[str]] `. A flattening process is required to convert it into `List[str]` to calculate the metrics.

Reply:
If `items` is a sequence of categories, then I think `sort` argument could be omitted since we need to calculate histogram ourselves.
Also, a more generic type hint for `items` might be `Union[Iterable[Hashable, ...], Iterable[Sequence[Hashable, ...], ...]]` in this case. What do you think?

Reply:
I agree. The `sort` argument could be omitted. The minimal requirements for metrics input should be compatible to:
```
- `List[]` like style
- `List[List[]]` like style
```
You can choose the proper type hint for these requirements.
In the future, we may support the metrics calculation groupby the users, but is not in a hurry.